lando/worker/cwlworkflow.py

```diff
@@ -21,82 +102,176 @@ def __init__(self, job_id, working_directory, output_directory, cwl_base_command
 
-    def run(self, cwl_file_url, workflow_object_name, job_order):
+    def run(self):
         """
-        Downloads the packed cwl workflow from cwl_file_url, runs it.
-        If cwl-runner doesn't exit with 0 raise JobStepFailed
-        :param cwl_file_url: str: url to workflow we will run (should be packed)
-        :param workflow_object_name: name of the object in our workflow to execute (typically '#main')
-        :param job_order: str: json string of input parameters for our workflow
+        Run job saving results in process_output, process_error_output, and return_code members.
+        :param command: [str]: array of strings representing a workflow command and its arguments
         """
-        job_order_filename = self._write_job_order_file(job_order)
-        workflow_file = os.path.join(self.working_directory, 'workflow.cwl')
-        urllib.urlretrieve(cwl_file_url, workflow_file)
-        if workflow_object_name:
-            workflow_file += workflow_object_name
-        local_output_directory = os.path.join(self.working_directory, self.output_directory)
-        command = self._build_command(local_output_directory, workflow_file, job_order_filename)
-        output, return_code = self.run_command(command)
-        if return_code != 0:
-            error_message = "CWL workflow failed with exit code: {}".format(return_code)
-            print(output)
-            raise JobStepFailed(error_message + output, output)
-
-    @staticmethod
-    def run_command(command):
-        """
-        Runs a command and saves the output by reading from pipes.
-        :param command: [str]: command to execute
-        :return: str, int: output from stderr/stdout and process exit status.
-        """
-        print(' '.join(command))
-        p = Popen(command, stdin=PIPE, stderr=PIPE, stdout=PIPE, bufsize=1)
-        output = ""
+        self.started = datetime.datetime.now()
+        p = Popen(self.command, stdin=PIPE, stderr=PIPE, stdout=PIPE, bufsize=1)
+        self.output = ""
+        self.error_output = ""
         while True:
             line = p.stderr.readline()
             if line:
-                output += line + "\n"
+                self.error_output += line + "\n"
             else:
                 break
         while True:
             line = p.stdout.readline()
             if line:
-                output += line + "\n"
+                self.output += line + "\n"
             else:
                 break
         p.wait()
-        return output, p.returncode
+        self.return_code = p.returncode
+        self.finished = datetime.datetime.now()
+
+    def total_runtime_str(self):
+        """
+        Returns a string describing how long the job took.
+        :return: str: "<number> minutes"
+        """
+        elapsed_seconds = (self.finished - self.started).total_seconds()
+        return "{} minutes".format(elapsed_seconds / 60)
+
+
+class ResultsDirectory(object):
```

Comment:
Agreed on user specifying output project and directory name.
Transferring the project also keeps the user from seeing partial data sets or uploads that failed to complete.